### ISSUE TITLE
Segmentation fault when running odp

### ISSUE THREAD

**Issue by bestgen** [Linaro/odp#592]
Hello. When I try to run a webserver with odp, it show the segmentation fault (core dumped). 
```
HW time counter freq: 3408006263 hz

_ishm.c:881:_odp_ishm_reserve():No huge pages, fall back to normal pages. check: /proc/sys/vm/nr_hugepages.
PKTIO: initialized loop interface.
PKTIO: initialized pcap interface.
PKTIO: initialized ipc interface.
PKTIO: initialized socket mmap, use export ODP_PKTIO_DISABLE_SOCKET_MMAP=1 to disable.
PKTIO: initialized socket mmsg,use export ODP_PKTIO_DISABLE_SOCKET_MMSG=1 to disable.

ODP system info
---------------
ODP API version: 1.17.0
CPU model:       Intel(R) Core(TM) i7-6700 CPU
CPU freq (hz):   3407880000
Cache line size: 64
Core count:      4

Running ODP appl: "ofp_netwrap"
-----------------
IF-count:        1
Using IFs:       ens33

/home/admin/Downloads/ofp/scripts/ofp_netwrap.sh: line 8:  7365 Segmentation fault      (core dumped) LD_PRELOAD=libofp_netwrap_crt.so.0.0.0:libofp.so.0.0.0:libofp_netwrap_proc.so.0.0.0 $@
```

I set up odp on the virtual machine based on Vmware, and I have allocate the virtual machine 4 cores and 2GB memory. 

Does anyone have some idea? Thanks.

**Comment by bogdanPricope** [comment]
Are you trying to run OFP example application webserver (https://github.com/OpenFastPath/ofp/tree/master/example/webserver)?

If so, don't use ofp_netwrap, as that was meant for running native Linux application (mainly some version of nginx) on top of OFP/ODP and not ODP or OFP/ODP applications.

Instead try this:

sudo www_dir=${SERVER_ROOT_DIR} ${BUILD_DIR}/ofp/example/webserver/webserver -i ${dev} -f ${OFP_CONFIG_FILE} -c ${OFP_CORES} &

where:
 - www_dir points to webserver root folder (index.html)
 - ${BUILD_DIR} is build directory
 - ${dev} is your device (ens33 in your case)
- ${OFP_CONFIG_FILE}  is a batch of CLI commands - you can set interface address, etc.
e.g.
loglevel set error
ifconfig fp0 192.168.100.1/24

 - ${OFP_CORES} is core count (all by default)

Note: You are using 'socket' pktio type, so packets are received through a raw socket -  You may need to make sure linux network stack is not also process your packets -  use iptables to block access to packets

sudo iptables -A FORWARD -i $dev -j DROP
sudo iptables -A INPUT -i $dev -j DROP

Take a look at this: https://git.linaro.org/lng/ofp_ci.git/tree/ofp_webserver/server/scripts/ofp_webserver.sh

**Comment by bestgen** [comment]
Hello @bogdanPricope  and thanks for you reply. 
As I understand,  your suggestion is to use the ofp webserver rather than nginx, as this example "starting-an-example-application":
https://github.com/OpenFastPath/ofp/blob/master/docs/ofp-user-guide.adoc#starting-an-example-application

If so, I have tried this example before and got the same problem. Actually, I got the same problem when I try to run all of the examples of ofp. Hence i just assume the problem may be due to the system resource limits, or something else? 

Here is the output when I try to start the ofp webserver. (^C is because the terminal has no response.)
```
[admin@localhost ofp]$ ./example/webserver/webserver -i ens33 -c 4 -f ./example/webserver/ofp.cli &
[1] 12353
[admin@localhost ofp]$ RLIMIT_CORE: 0/-1
Setting to max: 0
HW time counter freq: 3408002633 hz

_ishm.c:881:_odp_ishm_reserve():No huge pages, fall back to normal pages. check: /proc/sys/vm/nr_hugepages.
PKTIO: initialized loop interface.
PKTIO: initialized pcap interface.
PKTIO: initialized ipc interface.
PKTIO: initialized socket mmap, use export ODP_PKTIO_DISABLE_SOCKET_MMAP=1 to disable.
PKTIO: initialized socket mmsg,use export ODP_PKTIO_DISABLE_SOCKET_MMSG=1 to disable.

ODP system info
---------------
ODP API version: 1.17.0
CPU model:       Intel(R) Core(TM) i7-6700 CPU
CPU freq (hz):   3407839000
Cache line size: 64
Core count:      4

Running ODP appl: "webserver"
-----------------
IF-count:        1
Using IFs:       ens33

^C
[1]+  Segmentation fault      (core dumped) ./example/webserver/webserver -i ens33 -c 4 -f ./example/webserver/ofp.cli
```

**Comment by bogdanPricope** [comment]
First, you need to run it as superuser.
Second, yes is crashing in my case, too...

Thread 3 "webserver" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffaa5ff700 (LWP 1964)]
0x000000000040a6ee in default_event_dispatcher (arg=0x40a370 <ofp_eth_vlan_processing>) at ofp_pkt_processing.c:73
73              int rx_burst = global_param->evt_rx_burst_size;
(gdb) bt
#0  0x000000000040a6ee in default_event_dispatcher (arg=0x40a370 <ofp_eth_vlan_processing>) at ofp_pkt_processing.c:73
#1  0x00007ffff7bd32f7 in _odph_thread_run_start_routine (arg=0x7fffffffd4d0) at threads.c:56
#2  0x00007ffff6fa36ba in start_thread (arg=0x7fffaa5ff700) at pthread_create.c:333
#3  0x00007ffff6cd941d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
(gdb) print global_param
$1 = (ofp_global_param_t *) 0x0

It needs a simple fix -  I'll go bother our OFP guys ...

**Comment by bogdanPricope** [comment]
Use commit 79b2e986db5cecc9b11b4918c1c56d8f6a2c2b9b (I test it odp 1.19.0.0 but should work with 1.17 as well).

See progress in here: https://github.com/OpenFastPath/ofp/issues/203

**Comment by bogdanPricope** [comment]
BTW, what was OFP version (commit) in your test?

**Comment by sovu** [comment]
Issue was solved in OFP master at commit f7ba7a

**Comment by bestgen** [comment]
Sorry for the late reply and thanks for your works @bogdanPricope @sovu . Unfortunately, the updated codes still doesn't work for my test and has the exactly same problem as before. 

Second, I don't see any version information in the codes or on the OFP homepages. Anyway, I follow the guide (https://github.com/OpenFastPath/ofp#compiling-and-building-ofp) and get the newest codes through the github page of OFP.  @bogdanPricope 

```
[admin@localhost ofp]$ sudo ./example/webserver/webserver -i ens33 -c 4 -f ./example/webserver/ofp.cli &
[2] 61480
[admin@localhost ofp]$ RLIMIT_CORE: 0/-1
Setting to max: 0
HW time counter freq: 3408007402 hz

PKTIO: initialized loop interface.
PKTIO: initialized pcap interface.
PKTIO: initialized ipc interface.
PKTIO: initialized socket mmap, use export ODP_PKTIO_DISABLE_SOCKET_MMAP=1 to disable.
PKTIO: initialized socket mmsg,use export ODP_PKTIO_DISABLE_SOCKET_MMSG=1 to disable.

ODP system info
---------------
ODP API version: 1.17.0
CPU model:       Intel(R) Core(TM) i7-6700 CPU
CPU freq (hz):   3407839000
Cache line size: 64
Core count:      4

Running ODP appl: "webserver"
-----------------
IF-count:        1
Using IFs:       ens33

^C
[2]-  Segmentation fault      sudo ./example/webserver/webserver -i ens33 -c 4 -f ./example/webserver/ofp.cli
```

**Comment by bestgen** [comment]
I checked the codes and found that there might be some problems in ofp_init.c and ofp-arp_ck.c files.
For example, in ofp_init.c (line 257 to 287):

```
void ofp_init_global_param_from_file(ofp_global_param_t *params, const char *filename)
{
	memset(params, 0, sizeof(*params));
	params->pktin_mode = ODP_PKTIN_MODE_SCHED;
	params->pktout_mode = ODP_PKTIN_MODE_DIRECT;
	params->sched_sync = ODP_SCHED_SYNC_ATOMIC;
	params->sched_group = ODP_SCHED_GROUP_ALL;
#ifdef SP
	params->enable_nl_thread = 1;
#endif /* SP */
	params->arp.entries = OFP_ARP_ENTRIES;
	params->arp.hash_bits = OFP_ARP_HASH_BITS;
	params->arp.entry_timeout = OFP_ARP_ENTRY_TIMEOUT;
	params->arp.saved_pkt_timeout = OFP_ARP_SAVED_PKT_TIMEOUT;
	params->evt_rx_burst_size = OFP_EVT_RX_BURST_SIZE;
	params->pcb_tcp_max = OFP_NUM_PCB_TCP_MAX;
	params->pkt_pool.nb_pkts = SHM_PKT_POOL_NB_PKTS;
	params->pkt_pool.buffer_size = SHM_PKT_POOL_BUFFER_SIZE;
	params->pkt_tx_burst_size = OFP_PKT_TX_BURST_SIZE;
	params->num_vlan = OFP_NUM_VLAN;
	params->mtrie.routes = OFP_ROUTES;
	params->mtrie.table8_nodes = OFP_MTRIE_TABLE8_NODES;
	params->num_vrf = OFP_NUM_VRF;
	params->chksum_offload.ipv4_rx_ena = OFP_CHKSUM_OFFLOAD_IPV4_RX;
	params->chksum_offload.udp_rx_ena = OFP_CHKSUM_OFFLOAD_UDP_RX;
	params->chksum_offload.tcp_rx_ena = OFP_CHKSUM_OFFLOAD_TCP_RX;
	params->chksum_offload.ipv4_tx_ena = OFP_CHKSUM_OFFLOAD_IPV4_TX;
	params->chksum_offload.udp_tx_ena = OFP_CHKSUM_OFFLOAD_UDP_TX;
	params->chksum_offload.tcp_tx_ena = OFP_CHKSUM_OFFLOAD_TCP_TX;
	read_conf_file(params, filename);
}
```
nearly all of the fields,such as arp, chksum_offload, etc, cannot be resolved. So as ifx, ipv4_addr,etc, in ofp_arp_ck.c. There are hundreds of "could not be resolved" errors in total.

**Comment by bestgen** [comment]
I checked the codes and found many fields cannot be resolved.
First, in ofp_init.c files, there are many fields that cannot be resolved, for example:
static __thread struct ofp_global_config_mem *shm;

```
__thread ofp_global_param_t *global_param = NULL;

static void schedule_shutdown(void);
static void cleanup_pkt_queue(odp_queue_t pkt_queue);

static int ofp_global_config_alloc_shared_memory(void)
{
	shm = ofp_shared_memory_alloc(SHM_NAME_GLOBAL_CONFIG, sizeof(*shm));
	if (shm == NULL) {
		OFP_ERR("ofp_shared_memory_alloc failed");
		return -1;
	}
	global_param = &shm->global_param;
	return 0;
}
```
ofp_global_param_t, SHM_NAME_GLOBAL_CONFIG, and global_param cannot be solved in my case. Actually, nearly all of the "param-related" fields cannot be resolved, such as the line 259 to 285 in ofp_init.c. I think that is one of the reasons.

Second, in ofp_arp_ck.c files, there are many fields cannot be solved, too. I found there are three header files, <ck_epoch.h>, <ck_queue.h> <ck_pr.h>includes, while these three header files doesn't exist in the include files.
